### ISSUE TITLE
Uploading very large files

### ISSUE THREAD

**Issue by serge-wq** [Tewr/BlazorFileReader#162]
Hi, I'm using your component to read very large files (>1GB) and upload them to object storage. I am using server-side Blazor. I have configured the hub options `MaximumReceiveMessageSize` to 10Mb but I am using ´IFileReference.OpenReadAsync´ method to read the stream and the stream goes directly to the AmazonS3 transfer utility client. The application takes up to several hours in the uploading process (currently I'm up to two hours with no response). I get to the uploading instruction with the debugger so I think what's slowing the process is reading the Stream from the client to the server. There is no exception thrown  and everything is working correctly. It just seems that reading the file is taking incredibly long.

For reference I have tested the AmazonS3 client through an API, and using Postman for the requests a single file is uploaded in ~25 min. 

Are there any tips you can share or what would the best approach to accomplish this task?

**Comment by Tewr** [comment]
Hello. 

Short answer:

You could try to work around this by yourself by using multi-threading: Open multiple parallel file streams: like if you want to read bytes 0-20, read 0-5, 5-10, 10-15, 15-20 in parallel and then stich them back together. 

Long answer:
It's meticulous though, and typically that would be something a library like mine should solve.

First of all, my current implementation for server-side blazor is "naive", playing by "the rules". This has always been a "WASM-First" project. This imposes some limits on speed, for full details see @SteveSandersonMS <a href="http://blog.stevensanderson.com/2019/09/13/blazor-inputfile/">Blog post</a> for a more detailed explanation.

I have a <a href="https://github.com/Tewr/BlazorFileReader/pull/160">pull request</a> open working on a similar problem. Would probably adress this situation as well. I need to do some additional testing... Basically my pull request avoids the need for setting `MaximumReceiveMessageSize` and does that parallel stitching mentioned above. I think I might push it out in a few days if I have the time, but I havent decided if I should push is as an opt-in experimental feature, opt-out, or default behaviour. Its a similar approach to that of Steve, however, in the spirit of my library, it's focused on end-user control using the buffer size to control memory consumption vs speed - Steve's current implementation is fully focused on speed using full pre-fetching from the client at the cost of memory - maybe not ideal in your case.

**Comment by serge-wq** [comment]
Alright, I'm going to dig into what you are suggesting. I'm very new working with this kind of stuff. Regarding the feature opt-in or out mode, I can't see very clear what are the different implications for the webassembly devs in contrast to the server-side devs. Although, I like what you are saying about the focus on the end-user control of memory consumption vs speed.

Thanks a lot for taking the time to point me in the right direction. Great work with this library.

**Comment by Tewr** [comment]
I have pushed the change in the latest version. MaximumReceiveMessageSize is no longer neccessary to set. This should set you free to be able to use whatever buffersize you want, assuming you have control over the buffersize (this depends on the api of the "AmazonS3 transfer utility client").

basically, the higher the buffer size, you use more ram but transfer is faster. Up to a certain limit of course, exhausting your server ram won't do you any favors in terms of speed.

You could maybe post some of your code and I'll see if I can make any suggestions.

**Comment by serge-wq** [comment]
Sorry for the delayed response and thanks for the update.

I removed `MaximumReceiveMessageSize` from the configuration and set the AmazonS3 client buffer to 300MB, which I think would be fine for our server. I've been using `IFileReference.OpenReadAsync` and the async stream it generates to pass it directly to the AS3 client and performance has been lots better. I think the only thing now affecting is the client's internet connection which I have no control of. To get a better picture of how the application will behave I need a little bit more usage and testing. I can come and update when that happens. 

Something I noticed though, is that if I use `IFileReference.OpenReadAsync` instead of `IFileReference.CreateMemoryStreamAsync` to get an image stream, when I pass the stream to an ImageMagick object ImageMagick complains about not recognizing the image format. Any ideas about why this happens? Seems a little bit weird to me. Does OpenReadAsync changes the first bytes of the stream in any way? Or maybe the async stream is the reason?

**Comment by Tewr** [comment]
CreateMemoryStreamAsync uses OpenReadAsync and then reads it to the end, so it's the same data. However, not sure if imagemagic supports async-only streams. If imagemagic tries to call Read instead of ReadAsync on the stream, the stream will throw an exception, and that might be the underlying cause for the image format error, even though it sounds a bit odd. Did you not get an inner exception complaining about reading the stream sync?

**Comment by Tewr** [comment]
As per this issue https://github.com/dlemstra/Magick.NET/issues/422 might need to tell imagemagic what the file extension is so that the format can be guessed before reading the entire file

**Comment by serge-wq** [comment]
I did tried that, but even telling ImageMagick which file extension the image was it threw error that the format didn't match. I'm guessing the problem comes from the call to ReadAsync you mention, as I am not aware of async method inside the library. Any how, it was not a big problem for me to use CreateMemoryStreamAsync in this case because they were only a couple of MB in size.

**Comment by Tewr** [comment]
If the method you are calling in imagemagic is not async, it is indeed highly unlikely that it calls ReadAsync. Then basically CreateMemoryStreamAsync is your only choice in this case.

**Comment by serge-wq** [comment]
Hi! To give an update on this issue, everything has been working very well. I think the problem with the images mentioned above is due to ImageMagick.NET not being asynchronous. I am going to close this issue now as I think everything has been solved. Thanks a lot!
